### PR TITLE
Add COMUDF

### PR DESCRIPTION
**Pull request by matbryan52**
Adds `COMUDF`, continued from #1392 .

Notes:
- The default analysis parameters have been lifted into the definition of the `COMParams(NamedTuple)`, which makes the default functionality a bit clearer without digging into the code of `get_task_data`.
- Added a `@classmethod` `COMUDF.with_params(...)` to allow construction of a UDF with the desired `COMParams`, but mainly to give a clear place where the behaviour of each parameter is documented.
- Added a `raw_com` result buffer for the position of the CoM in the coordinate system of the frame - i.e. COMUDF returns the CoM, in addition to the *shift* in the CoM + field-derived results.
- Documented all of the result buffers in the docstring of `COMUDF`, again for visibility.

Two ideas:
- Any enthusiasm for renaming the UDF `CoMUDF` ? It would make it feel like less of a global variable, to me.
- I've always wondered about implementing a mode where the `raw_shifts` (and hence field++) are calculated from the *mean* of `raw_com` result, so that any error in the specification of `cy`/`cx` relative to the actual data centre doesn't give shifts / field which are all offset in one direction. Could also return the mean value and/or reference value in a result-only buffer for transparency. Any thoughts on that?

* [x] Check if this is still compatible with the `iCoM` repo

## Contributor Checklist:

* [x] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [x] I have added/updated documentation for all user-facing changes
* [x] I have added/updated test cases
* [ ] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if changes were made to the GUI)

## Reviewer Checklist:

* [ ] `/azp run libertem.libertem-data` passed
* [ ] No import of GPL code from MIT code

### CODE REVIEW

**Comment by matbryan52** [comment]
> Hi @matbryan52, :+1: for renaming! Would you mind if I take over the PR to finish it for the 0.12 release, or what are your plans with it? :-)

Go ahead! I'm still away this week so won't be able to work on it until the next. Feel free to push onto my branch if that's helpful.

**Comment by uellue** [comment]
> * I've always wondered about implementing a mode where the `raw_shifts` (and hence field++) are calculated from the _mean_ of `raw_com` result, so that any error in the specification of `cy`/`cx` relative to the actual data centre doesn't give shifts / field which are all offset in one direction. Could also return the mean value and/or reference value in a result-only buffer for transparency. Any thoughts on that?

The correct compensation would be subtracting a vacuum reference. Subtracting the mean works if there's no long-range field. If we pass the beam through a constant electric or magnetic field, an equal shift of all points could be the actual result. On the other hand, I agree that the shift relative to the mean or even relative to a best-fit linear gradient (descan error) is often a relevant value. :+1: for additional result-only buffers!

**Comment by uellue** [comment]
After discussion with @sk1p: Correct the field instead of adding more result buffers so that iCoMUDF works organically with the flat field correction, where it is probably particularly useful.

File: src/libertem/udf/com.py [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+        non-NaN values even in :code:`ROI == False` pixels due to the behaviour
+        of :code:`np.gradient`.
+
+    Parameters
+    ----------
+    com_params : CoMParams
+        A :class:`CoMParams` instance containing the parameters
+        for this UDF. By default will create a CoMParams instance
+        which performs whole-frame CoM with results in the coordinates
+        of the frame # CHECKTHIS
```
Comment (sk1p) [inline src/libertem/udf/com.py:306]: Aren't the results relative to the detector center? See `get_params` below.

File: src/libertem/udf/com.py [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+
+    Examples
+    --------
+    >>> udf = CoMUDF()
+    >>> result = ctx.run_udf(dataset=dataset, udf=udf)
+    >>> result["magnitude"].data.shape
+    (16, 16)
+    """
```
Comment (sk1p) [inline src/libertem/udf/com.py:314]: Hmm, could we have more examples here? One that uses `with_params` maybe?

File: src/libertem/udf/com.py [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+        :meth:`apply_correction`
+    roi : Optional[numpy.ndarray]
+        Selector for values to consider in the statistics, compatible
+        with indexing an array with the shape of y_centers and x_centers.
+        By default, everything except the last row and last column are used
+        since these contain artefacts.
+
+    Returns
+    -------
+    GuessResult : relative to current values
```
Comment (sk1p) [inline src/libertem/udf/com.py:211]: Can we have an example here how to pipe the result from the `CoMUDF` into `guess_corrections`? I think that would be a common use-case.

That would also mean this function needs to be added to the reference docs, or it needs to be exposed (as a static method?) on the `CoMUDF` itself.

File: src/libertem/udf/com.py [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+            cy=cy, cx=cx, r=r, ri=ri, scan_rotation=scan_rotation, flip_y=flip_y,
+            regression=regression,
+        )
+        return cp
+
+    def get_task_data(self):
+        sig_shape = tuple(self.meta.dataset_shape.sig)
+        com_params = self.get_params()
+        if len(sig_shape) != 2:
+            raise ValueError('CoMUDF only works with 2D sig shape.')
```
Comment (sk1p) [inline src/libertem/udf/com.py:458]: Should we have test cases for incompatible sig/nav shapes? Should be easy enough with a memory dataset.

File: src/libertem/udf/com.py [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+
+            Please note that a regression calculated on the data itself may
+            distort valid results, such as the effect of long-range fields.
+            Furthermore, it may yield erronerous results due to aliasing on
+            periodic structures or boundary effects. Subtracting a reference
+            measurement or regression on a reference measurement instead of a
+            regression on the data may give more reliable results.
+        """
+        if ri >= r:
+            raise ValueError('Inner radius must be less than outer radius for annular CoM')
```
Comment (sk1p) [inline src/libertem/udf/com.py:391]: Could we have a test case for this edge case, too?

File: src/libertem/udf/masks.py [hunk 6]
```diff
@@ -1,11 +1,99 @@
...
-from libertem.udf import UDF
+from libertem.udf import UDF, UDFMeta
 from libertem.common.container import MaskContainer
 from libertem.common.numba import rmatmul
 
 
+class ApplyMasksEngine:
+    def __init__(self, masks: MaskContainer, meta: UDFMeta, use_torch: bool):
+        self.masks = masks
+        self.meta = meta
```
Comment (sk1p) [inline src/libertem/udf/masks.py:12]: Just a comment, nothing to fix right now: taking a `UDFMeta` instance here means this class is right now only usable in UDFs. I do really love that this code is now reusable, and in the future we might want to lift it up further to make it usable stand-alone (possibly by extracting the relevant parameters into our own parameter class, which is then easily constructable from ` UDFMeta`.

**Review by sk1p — CHANGES_REQUESTED** [review]
Looks almost ready! I've added some comments inline, most of them are just about minor changes, and adding test cases for the edge cases (to make sure they continue to be checked properly in the future)

About the CI failure, a rebase to current `master` should fix the flakyness of the affected test case, by adjusting the timings to those used in the integration tests running at UDF level, which were running successfully since #1441 was merged, as far as I can see.

I've opened #1482 for the remaining work to surface the new features in the web GUI.

**Comment by matbryan52** [comment]
@uellue I'm happy to cover the review comments in the morning as part of looking at the PR in detail - I think some of them concern bits of docstring that I myself wrote.

File: src/libertem/udf/com.py (hunk 3 above)
Comment (matbryan52) [reply src/libertem/udf/com.py:211]: Agreed for the example, even I am not 100% sure how to pipe from one to the other...

The `guess_corrections` function is already documented under `analysis.com`, though its docstring references the functions `center_shifts` and `apply_corrections` which are themselves undocumented, currently. I will see if I can write something complete with the new functionality offered with CoMUDF.

Comment (matbryan52) [reply src/libertem/udf/com.py:211]: Do you have a good way to :autodoc: a NamedTuple or Enum (e.g. https://libertem.github.io/LiberTEM/reference/api.html#libertem.analysis.com.GuessResult ) ? I think there are Sphinx extension that can do it, but otherwise the output is quite unhelpful!

Comment (sk1p) [reply src/libertem/udf/com.py]: Yeah, named tuples aren't properly supported by sphinx, yet: https://github.com/sphinx-doc/sphinx/issues/8717

I wouldn't be opposed to using [`autonamedtuple`](https://sphinx-toolbox.readthedocs.io/en/latest/extensions/more_autodoc/autonamedtuple.html) from sphinx-toolbox, if that is compatible with the rest of our sphinx extensions.

**Comment by matbryan52** [comment]
@uellue Thanks for picking up the change to sphinx, seems like this is nearly ready to go?

I think you noticed that I added an `IntEnum` for `RegressionOptions`, hopefully that's an OK choice @sk1p  ? We don't have that pattern elsewhere right now, I think.

**Comment by matbryan52** [comment]
Also an observation - we import and document `guess_corrections` under `libertem.analysis.com`, which means the example in the docstring doesn't actually match the documentation of that file (analysis api rather than udf api). We could choose to remove the `guess_corrections` and `GuessResult` members documented here:

https://github.com/LiberTEM/LiberTEM/blob/060613939719efbe3847b053f893a250bcb3984d/docs/source/reference/api.rst?plain=1#L21-L22

**Comment by sk1p** [comment]
> @uellue Thanks for picking up the change to sphinx, seems like this is nearly ready to go?

Yup, should be ready to go if CI is green.

> I think you noticed that I added an `IntEnum` for `RegressionOptions`, hopefully that's an OK choice @sk1p ? We don't have that pattern elsewhere right now, I think.

Yes, `IntEnum` is :+1:. I'd also be fine with a regular enum but it doesn't really matter to me (int enum has benefit that you don't necessarily need to import the specific enum itself, if you know the constants, with the readability drawback that comes with it...)

**Comment by uellue** [comment]
> @uellue Thanks for picking up the change to sphinx, seems like this is nearly ready to go?

Oh, sry I hope I didn't disrupt your work by pushing?

> I think you noticed that I added an `IntEnum` for `RegressionOptions`, hopefully that's an OK choice @sk1p ? We don't have that pattern elsewhere right now, I think.

With autoenum the documentation for that one looks pretty decent, I like it!

**Comment by matbryan52** [comment]
> > @uellue Thanks for picking up the change to sphinx, seems like this is nearly ready to go?
> 
> Oh, sry I hope I didn't disrupt your work by pushing?

Not at all, it had slipped my mind to do that last step adding `sphinx_toolbox`.

**Comment by sk1p** [comment]
> Also an observation - we import and document `guess_corrections` under `libertem.analysis.com`, which means the example in the docstring doesn't actually match the documentation of that file (analysis api rather than udf api). We could choose to remove the `guess_corrections` and `GuessResult` members documented here:

Yeah, I'd be fine with documenting the new import location only, the old ones are basically only in place for backwards-compatibility.
